### PR TITLE
feat: [MR-523] Prevent enqueueing multiple responses for the same callback

### PR DESCRIPTION
**Pull request by alin-at-dfinity**
[[MR-523]] Currently, when inducting a response we only check that it matches an existing callback, dropping any mismatched responses and raising a critical error. However, a malicious subnet (or a bug) could produce a second response for the same callback before the first one has been executed (and the callback closed). This second response would also get enqueued and consume an extra slot reservation; which would later result in dropping a legitimate response due to the queue no longer having a response slot reservation for it.

We maintain an explicit, transient set of callback IDs whose responses are already enqueued in input queues; and check against this set for potential duplicate responses. If a duplicate is detected, we return an error if this is a guaranteed response (which results in a critical error); but silently ignore duplicate best-effort responses (as this may be a response delivered after we have already produced a local timeout reject response).

This functionality is required in order to safely generate local timeout reject responses for expired best-effort callbacks.

[MR-523]: https://dfinity.atlassian.net/browse/MR-523?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: rs/pocket_ic_server/BUILD.bazel [hunk 1]
```diff
@@ -55,10 +55,9 @@ LIB_DEPENDENCIES = [
     "@crate_index//:http",
     "@crate_index//:http-body-util",
     "@crate_index//:hyper",
+    "@crate_index//:hyper-rustls",
```
[the remark this answers is not in the document]
Comment (alin-at-dfinity) [reply rs/pocket_ic_server/BUILD.bazel]: Yeah, sorry. I was fighting a test failure and was apparently to eager to get the fix onto this branch.

**Comment by derlerd-dfinity** [comment]
> if due to malicious behavior

Nit: I would make it clear in the PR description that this requires a malicious subnet to become an issue. As things are written currently one might suspect that this could become a problem even without any subnet being dishonest.

I will do a more in-depth review later, but I have additional questions regarding the new invariant:
1. Are we sure no two callbacks in the same queue should be an invariant (i.e., is it really worth crashing the subnet whenever it is violated)? It seems execution would survive just fine if the same callback ID is enqueued into a queue twice, so a critical error if this situation is ever encountered on mainnet might be fine as well?
2. If the answer to 1 is that we are sure that we want an invariant, how can we be sure that it holds today on every mainnet subnet and this change is safe to roll out?

**Comment by alin-at-dfinity** [comment]
I updated the PR description, to explicitly say "malicious subnet" instead of merely "malicious behavior". (Do note that a bug in Message Routing could technically also result in a duplicate response, without active malicious intent.)

To answer your questions:
1. Two responses for the same callback is problematic. Not due to Execution, which will nicely ignore a response to a non-existent callback (because what else could it do, apart from panicking?). But due to response slot reservations: if we enqueued two responses for one callback, by definition we will have to eventually drop a response to a different callback (because its slot reservation will be gone). Meaning a hanging callback.

   We could, of course, relax the invariant on reservation slots (i.e. accept a response even if we don't have a reservation for it; and simply record a critical error at that point) but that seems like an unsafe alternative: by the time we would get paged and start doing something about it, the subnet could already be flooded with hundreds of GB worth of duplicate responses. I would rather have this be a hard invariant and risk a crash (and straightforward subnet recovery) in the unlikely case of a bug.

2. We cannot be 100% sure that this will hold at the precise moment when we upgrade each subnet. But until now, we haven't seen any responses dropped because their slot reservations were gone, which is a pretty strong indication that this has never happened.

**Comment by derlerd-dfinity** [comment]
Thanks for your response. 

> [...] if we enqueued two responses for one callback, by definition we will have to eventually drop a response to a different callback (because its slot reservation will be gone). Meaning a hanging callback. [...]

I guess this is a good argument, as going ahead with this violated would impact correctness. So no objections against the invariant from my side. We should make sure that we are convinced that the invariant holds today on mainnet before we start explicitly enforcing it to avoid crashes on rollout. 

I'll review the PR in more depth tomorrow.

File: rs/replicated_state/src/canister_state/queues.rs [hunk 2]
```diff
@@ -393,18 +410,32 @@ impl CanisterQueues {
...
+                        if !self
+                            .callbacks_with_enqueued_response
+                            .insert(response.originator_reply_callback)
+                        {
+                            debug_assert_eq!(Ok(()), self.test_invariants());
+                            // This is a critical error for guaranteed responses.
+                            if response.deadline == NO_DEADLINE {
+                                return Err((
+                                    non_matching_response("Duplicate response", response),
+                                    msg,
```
Comment (derlerd-dfinity) [inline rs/replicated_state/src/canister_state/queues.rs:424]: Just to double check: returning an error here will be handled upstream by raising a critical error but _not_ crashing right?

**Review by derlerd-dfinity — COMMENTED** [review]
Generally the change looks good to me. My only concern is that we now have various places where we have to add/remove entries to keep up the invariant which makes it harder to convince oneself that the invariant holds, and might also make future changes more difficult. Can you think of other ways that would better encapsulate this invariant?

File: rs/replicated_state/src/canister_state/queues.rs (hunk 2 above)
Comment (stiegerc) [reply rs/replicated_state/src/canister_state/queues.rs:424]: There is no exception for just this error type so it doesn't add any panics that were not there before.

File: rs/replicated_state/src/canister_state/queues/tests.rs [hunk 3]
```diff
@@ -1752,33 +1895,41 @@ fn test_stats_best_effort() {
...
     queues.output_into_iter().next().unwrap();
 
     // Actually enqueue the messages.
     queues
-        .push_input(request.clone().into(), LocalSubnet)
+        .push_input(request3.clone().into(), LocalSubnet)
         .unwrap();
     queues
-        .push_input(response.clone().into(), LocalSubnet)
+        .push_input(response2.clone().into(), LocalSubnet)
```
[the remark this answers is not in the document]
Comment (alin-at-dfinity) [reply rs/replicated_state/src/canister_state/queues/tests.rs:1936]: The reason why I'm pushing a request then a response in the input queue; and the reverse in the output queue; has to do with my wanting to mix things up when popping the messages: had I popped two requests and been left with two responses, I imagine some subtle bugs could have slipped by (such as decrementing the inbound request count for the outbound request and the other way around).

I could also inline making the reservations, but I was aiming to have a single, concise block that documents exactly what is enqueued into `queues` at the end of the set up phase.

File: rs/replicated_state/src/canister_state/queues/tests.rs [hunk 4]
```diff
@@ -1864,33 +2015,41 @@ fn test_stats_guaranteed_response() {
...
         .unwrap();
```
[the remark this answers is not in the document]
Comment (alin-at-dfinity) [reply rs/replicated_state/src/canister_state/queues/tests.rs:2057]: As above, I was aiming to have a single block of code where all the messages actually in `queues` are being enqueued, so it's easy to eyeball withoug having to skip irrelevant bits (such as repeated pushing and popping of requests).

File: rs/replicated_state/src/canister_state/queues/tests.rs [hunk 5]
```diff
@@ -1744,6 +1841,52 @@ fn decode_backward_compatibility() {
...
+#[test]
+fn decode_duplicate_inbound_response() {
+    let mut queues = CanisterQueues::default();
+
+    // Make 2 input queue reservations.
+    queues
+        .push_output_request(request(1, NO_DEADLINE).into(), UNIX_EPOCH)
+        .unwrap();
+    queues
+        .push_output_request(request(2, coarse_time(1)).into(), UNIX_EPOCH)
```
[the remark this answers is not in the document]
Comment (alin-at-dfinity) [reply rs/replicated_state/src/canister_state/queues/tests.rs]: Great suggestion, thanks. I guess that `SOME_DEADLINE` was a late addition, so it only made it into one test case.

Made `SOME_DEADLINE` a module level constant and used it everywhere the actual deadline made no difference (there are still some tests that need to time out some messages but not others, that use `coarse_time()` directly).

File: rs/replicated_state/src/canister_state/queues/tests.rs [hunk 6]
```diff
@@ -1744,6 +1841,52 @@ fn decode_backward_compatibility() {
...
+        encoded.clone(),
+        &StrictMetrics as &dyn CheckpointLoadingMetrics,
+    )
+        .try_into()
+        .unwrap();
+    assert_eq!(queues, decoded);
+
+    // Tweak the encoded queues so both responses have the same `CallbackId`.
+    for entry in &mut encoded.pool.as_mut().unwrap().messages {
+        let message = entry.message.as_mut().unwrap().r.as_mut().unwrap();
```
[the remark this answers is not in the document]
Comment (stiegerc) [reply rs/replicated_state/src/canister_state/queues/tests.rs:1885]: Ack. Thanks for clarifying.

File: rs/replicated_state/src/canister_state/queues/tests.rs [hunk 7]
```diff
@@ -1752,33 +1895,41 @@ fn test_stats_best_effort() {
...
-    let request = request(coarse_time(10));
-    let request_size_bytes = request.count_bytes();
-    let response = response_with_payload(1000, coarse_time(20));
-    let response_size_bytes = response.count_bytes();
+    // Best-effort requests and best-effort responses, to be enqueued one each into
+    // an input and an output queue.
+    let t10 = coarse_time(10);
+    let t20 = coarse_time(20);
+    let request1_ = request(1, t10);
+    let request2_ = request(2, t10);
```
[the remark this answers is not in the document]
Comment (stiegerc) [reply rs/replicated_state/src/canister_state/queues/tests.rs:1911]: I was guessing this but also curious. Fine by me.

File: rs/replicated_state/src/canister_state/queues/tests.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (stiegerc) [reply rs/replicated_state/src/canister_state/queues/tests.rs:1936]: There is no such ordering guarantee on the IC so I suppose it makes sense not to impose such patterns in tests. It was a bit confusing to read though but nvm.

File: rs/replicated_state/src/canister_state/queues.rs [hunk 8]
```diff
@@ -144,6 +144,12 @@ pub struct CanisterQueues {
     /// and remote subnet senders; as well as within the local subnet senders and
     /// remote subnet senders groups.
     input_schedule: InputSchedule,
+
```
Comment (derlerd-dfinity) [inline rs/replicated_state/src/canister_state/queues.rs:147]: GitHub prevents me from commenting on the actual line, but I think the comment regarding the invariants on `CanisterQueues` above may need updating as well.

**Review by derlerd-dfinity — APPROVED** [review]
Thank you. LGTM.

You may want to consider waiting with merging until the next release candidate is cut (presumably today) to avoid too many changes in one release. No strong opinion though -- leaving this up to you.

File: rs/replicated_state/src/canister_state/queues/tests.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (alin-at-dfinity) [reply rs/replicated_state/src/canister_state/queues/tests.rs:1936]: > There is no such ordering guarantee on the IC

What ordering guarantee are you referring to? If I enqueue a request, then a response into an input queue; and a response, then a request into an output queue; and I pop one input and one output; I am guaranteed to be left with a response in the input queue and a request in the output queue, no?

(Sure, technically there is no ordering guarantee between requests and responses in the spec, but there very much is in the implementation.)

[the remark this answers is not in the document]
Comment (stiegerc) [reply rs/replicated_state/src/canister_state/queues/tests.rs:1936]: I mean we guarantee the delivery of requests in the order they were made, responses may come in any order. So the way you made it here makes sense in that it doesn't introduce such a structure which could hide subtle bugs.

Anyways, not strong feelings one way or the other.

File: rs/replicated_state/src/canister_state/queues.rs (hunk 8 above)
Comment (alin-at-dfinity) [reply rs/replicated_state/src/canister_state/queues.rs:150]: Listed `callbacks_with_enqueued_response` as a hard invariant, thanks for pointing it out.

**Pull request merged**